### PR TITLE
More instances and fix faulty `TxFeePolicy` instance

### PR DESCRIPTION
**Pull request by adinapoli-iohk**
This small PR adds some instances (list below) and correct a faulty instance for `TxFeePolicy`. The problem, as has been rightly pointed out, was that it was not really resilient to change. The culprit was _not_ serialising everything as a `ByteString` blob, so that we could "swap container type" with no harm.

The same underlying idea has been crystallised in the `Test` module using a type `U` mimicking the real scenario where we need to swap between different representation of the type whilst preserving it. All the credits goes to @arybczak for coming up with the idea!

I have indeed verified, armed with this new `Property`, that my previous instance was faulty and that this new one is indeed sound. I love QC! 😉 

Instances added in this PR:

- [x] StakeDistribution
- [x] ApplicationName
- [x] SoftwareVersion
- [x] BlockVersion
- [x] Commitment
- [x] CommitmentsMap
- [x] VssCertificate
- [x] Opening
- [x] GtPayload
- [x] GtProof
- [x] (DataMsg MCCommitment)
- [x] (DataMsg MCOpening)
- [x] (DataMsg MCShares)
- [x] (DataMsg MCVssCertificate)
- [x] DHTKey
- [x] DHTData
- [x] MessageName
- [x] MsgGetHeaders
- [x] MsgGetBlocks
- [x] HandlerSpec
- [x] VerInfo
- [x] DlgPayload
- [x] Attributes () **(Just a placeholder waiting for the rework)**
- [x] EpochSlottingData
- [x] SlottingData
- [x] (DataMsg (UpdateProposal, [UpdateVote]))
- [x] (DataMsg UpdateVote)
- [x] (DataMsg ProxySKLight)
- [x] (DataMsg ProxySKHeavy)
- [x] (DataMsg ProxySKLightConfirmation)
- [x] SystemTag
- [x] UpdateVote
- [x] UpdateData
- [x] BlockVersionModifier
- [x] UpdateProposal
- [x] UpdateProposalToSign
- [x] UpdatePayload
- [x] VoteState
- [x] (U.PrevValue a)
- [x] USUndo
- [x] UpsExtra
- [x] DpsExtra
- [x] UndecidedProposalState
- [x] DecidedProposalState
- [x] ProposalState
- [x] ConfirmedProposalState
- [x] BlockVersionState
- [x] Address **(Just stubbed as Andrzej is actively working on it)**
- [x] (MerkleTree a)
- [x] (MerkleRoot a)
- [x] TxIn
- [x] TxOut
- [x] TxOutAux
- [x] Tx
- [x] TxInWitness
- [x] TxDistribution
- [x] TxSigData
- [x] TxAux
- [x] TxProof
- [x] TxPayload
- [x] UserSecret
- [x] WalletUserSecret
- [x] BackupPhrase
- [x] (DataMsg TxMsgContents)

cc @dcoutts

### CODE REVIEW

**Comment by arybczak** [comment]
Thanks. Hmm, now that I think of it, when we use encodeListLen, in deserialization we should use decodeListLenOf (when length is static) or decode it and then when e.g. we deserialize multiple constructors, after we decode tag and decide what constructor we're working with, we should check that the length is appropriate.

**Comment by adinapoli-iohk** [comment]
Hey!

I think this is already the case for most instances, for which I'm using this very simply combinator called [enforceSize](https://github.com/arybczak/cardano-sl/blob/cbor-migration/core/Pos/Binary/Cbor/Class.hs#L41-L46). I don't know if this is enough and/or using `decodeListLenOf` would be more appropriate here (hopefully the change is trivial although pervasive).

> after we decode tag and decide what constructor we're working with, we should check that the length is appropriate.

Yes, I was wondering about this very specific case, and I couldn't see the wood for the trees. The problem is that, in order to have a valid flat term encoding, you need to call `encodeListLen` _prior_ to actually encoding the tag. Doing this the other way around won't work, obviously.

Originally I thought this would be a dealbreaker as there was no way to know _a priori_ the final length as that is specific for each constructor, but now that I think more about this with you, I think we can get away with something like this (pseudo-code) both on manual instances as well as the TH ones:

``` haskell
decode = do
  len <- decodeListLen
  tag <- decode @Word8
  case tag of
    0 -> -- at this point we know how many arguments we expect next, and we can enforce the length 
    .. ->
```

In a nutshell, read the length straight away, but reuse it later.
I didn't do this originally (for some obscure reason) but it should be totally possible. The TH code will need to be amended as well, but hopefully it's not _that_ painful. At the moment I'm simply discarding the length information, but ideally we should use it for greater good.
